File: apps/burner-mac/Tests/FlagshipBurnerTests/VMConfigTests.swift
import XCTest
@testable import FlagshipBurnerCore

final class VMConfigTests: XCTestCase {

    // MARK: - Fixtures

    private func makeRecipe(bootUnlockMode: String? = nil,
                            diskEncryption: String? = nil) -> Recipe {
        let ac = RecipeAuthCode(version: 1,
                                serial: "01VMTEST",
                                username: "harry",
                                serverName: "home",
                                serverDomain: "home.harry.flagship.services",
                                delegatedPubKeyHex: String(repeating: "13", count: 32),
                                userPubKeyHex: String(repeating: "ea", count: 32),
                                issuedAt: 1_899_996_400_000,
                                expiresAt: 1_900_000_000_000)
        return Recipe(version: 2,
                      serverDomain: "home.harry.flagship.services",
                      username: "harry",
                      serverName: "home",
                      phoneDelegatedPubKeyHex: String(repeating: "13", count: 32),
                      registrationUrl: "https://flagship.services/api/server/register",
                      authCode: ac,
                      authCodeUserSignatureHex: String(repeating: "00", count: 64),
                      installerGitRef: "main",
                      rckPubKeyHex: String(repeating: "fd", count: 32),
                      blobSignatureHex: String(repeating: "f1", count: 64),
                      bootUnlockMode: bootUnlockMode,
                      diskEncryption: diskEncryption)
    }

    private func json(_ obj: [String: Any]) -> Data {
        try! JSONSerialization.data(withJSONObject: obj)
    }

    private let host16 = HostResources(cpuCount: 8, memoryBytes: 16 * VMResourcePlan.gib)

    // MARK: - Determinism + shape

    func testPlanIsDeterministic() {
        let recipe = makeRecipe()
        let raw = json(["version": 2])
        let a = VMConfig.plan(recipe: recipe, recipeJSON: raw, host: host16)
        let b = VMConfig.plan(recipe: recipe, recipeJSON: raw, host: host16)
        XCTAssertEqual(a, b)
    }

    func testPlanCarriesTheServerIdentityAndResources() {
        let cfg = VMConfig.plan(recipe: makeRecipe(), recipeJSON: json([:]), host: host16)
        XCTAssertEqual(cfg.name, "home.harry.flagship.services")
        XCTAssertEqual(cfg.serverDomain, "home.harry.flagship.services")
        XCTAssertEqual(cfg.username, "harry")
        XCTAssertEqual(cfg.serverName, "home")
        XCTAssertEqual(cfg.cpuCount, VMResourcePlan.vmCPUCount(host: host16))
        XCTAssertEqual(cfg.memoryBytes, VMResourcePlan.vmMemoryBytes(host: host16))
        XCTAssertEqual(cfg.mainDiskSizeBytes, VMResourcePlan.defaultMainDiskSizeBytes)
        XCTAssertEqual(cfg.networkMode, .nat)
    }

    // MARK: - Serial console ⇔ debug grant (the hard guardrail)

    func testProductionRecipeGetsNoSerialConsole() {
        let cfg = VMConfig.plan(recipe: makeRecipe(), recipeJSON: json(["version": 2]), host: host16)
        XCTAssertFalse(cfg.serialConsoleEnabled)
    }

    func testDebugGrantSiblingEnablesTheSerialConsole() {
        let raw = json(["version": 2, "debugGrant": "{\"grant\":{},\"signatureHex\":\"ab\"}"])
        let cfg = VMConfig.plan(recipe: makeRecipe(), recipeJSON: raw, host: host16)
        XCTAssertTrue(cfg.serialConsoleEnabled)
    }

    func testDebugGrantInsideTheEnvelopeShapeEnablesTheConsole() {
        // The sibling rides at the TOP level of the issued envelope, beside
        // `blob`/`blobSignature` — exactly where the canonical engine reads it.
        let raw = json(["blob": ["version": 2], "blobSignature": "f1",
                        "debugGrant": ["grant": ["issuedAt": 1], "signatureHex": "ab"]])
        let cfg = VMConfig.plan(recipe: makeRecipe(), recipeJSON: raw, host: host16)
        XCTAssertTrue(cfg.serialConsoleEnabled)
    }

    func testEmptyDebugGrantStringDoesNotEnableTheConsole() {
        // Mirrors the engine's asStr: an empty string is "absent".
        let raw = json(["version": 2, "debugGrant": ""])
        let cfg = VMConfig.plan(recipe: makeRecipe(), recipeJSON: raw, host: host16)
        XCTAssertFalse(cfg.serialConsoleEnabled)
    }

    // MARK: - Unlock policy from the signed blob

    func testEncryptedGuestAwaitsPhoneUnlockAtBoot() {
        // Absent diskEncryption ⇒ LUKS ⇒ boots into the sealed state.
        let cfg = VMConfig.plan(recipe: makeRecipe(), recipeJSON: json([:]), host: host16)
        XCTAssertTrue(cfg.diskEncrypted)
        XCTAssertTrue(cfg.awaitsPhoneUnlockAtBoot)
        XCTAssertEqual(cfg.bootUnlockMode, "auto")
    }

    func testApproveModeIsCarriedThrough() {
        let cfg = VMConfig.plan(recipe: makeRecipe(bootUnlockMode: "approve"),
                                recipeJSON: json([:]), host: host16)
        XCTAssertEqual(cfg.bootUnlockMode, "approve")
        XCTAssertTrue(cfg.awaitsPhoneUnlockAtBoot)
    }

    func testUnencryptedGuestBootsStraightThrough() {
        let cfg = VMConfig.plan(recipe: makeRecipe(diskEncryption: "none"),
                                recipeJSON: json([:]), host: host16)
        XCTAssertFalse(cfg.diskEncrypted)
        XCTAssertFalse(cfg.awaitsPhoneUnlockAtBoot)
    }

    // MARK: - Codable round-trip (persisted in config.json)

    func testCodableRoundTrip() throws {
        let raw = json(["debugGrant": "x"])
        let cfg = VMConfig.plan(recipe: makeRecipe(bootUnlockMode: "approve"),
                                recipeJSON: raw, host: host16)
        let data = try VMInventoryStore.encoder().encode(cfg)
        let back = try VMInventoryStore.decoder().decode(VMConfig.self, from: data)
        XCTAssertEqual(back, cfg)
    }
}

File: apps/builder-windows/tests/VMConfigTests.cs
using System;
using System.Text;
using System.Text.Json;
using Xunit;
using Flagship.Builder;
using Flagship.Builder.VM;

namespace Flagship.Builder.Tests;

/// <summary>Direct port of apps/builder-mac VMConfigTests.swift.</summary>
public class VMConfigTests
{
    // ---- Fixtures ----

    private static Recipe MakeRecipe(string? bootUnlockMode = null, string? diskEncryption = null)
        => new()
        {
            Version = 2,
            ServerDomain = "home.harry.flagship.services",
            Username = "harry",
            ServerName = "home",
            PhoneDelegatedPubKeyHex = new string('1', 64),
            RegistrationUrl = "https://flagship.services/api/server/register",
            AuthCode = new RecipeAuthCode
            {
                Version = 1,
                Serial = "01VMTEST",
                Username = "harry",
                ServerName = "home",
                ServerDomain = "home.harry.flagship.services",
                DelegatedPubKeyHex = new string('1', 64),
                UserPubKeyHex = new string('e', 64),
                IssuedAt = 1_899_996_400_000,
                ExpiresAt = 1_900_000_000_000,
            },
            AuthCodeUserSignatureHex = new string('0', 128),
            InstallerGitRef = "main",
            RckPubKeyHex = new string('f', 64),
            BlobSignatureHex = new string('f', 128),
            BootUnlockMode = bootUnlockMode,
            DiskEncryption = diskEncryption,
        };

    private static byte[] Json(string s) => Encoding.UTF8.GetBytes(s);

    private static readonly HostResources Host16 = new(8, 16 * VMResourcePlan.GiB);

    // ---- Determinism + shape ----

    [Fact]
    public void PlanIsDeterministic()
    {
        var recipe = MakeRecipe();
        var raw = Json("{\"version\":2}");
        var a = VMConfig.Plan(recipe, raw, Host16);
        var b = VMConfig.Plan(recipe, raw, Host16);
        Assert.Equal(a, b);
    }

    [Fact]
    public void PlanCarriesTheServerIdentityAndResources()
    {
        var cfg = VMConfig.Plan(MakeRecipe(), Json("{}"), Host16);
        Assert.Equal("home.harry.flagship.services", cfg.Name);
        Assert.Equal("home.harry.flagship.services", cfg.ServerDomain);
        Assert.Equal("harry", cfg.Username);
        Assert.Equal("home", cfg.ServerName);
        Assert.Equal(VMResourcePlan.VmCpuCount(Host16), cfg.CpuCount);
        Assert.Equal(VMResourcePlan.VmMemoryBytes(Host16), cfg.MemoryBytes);
        Assert.Equal(VMResourcePlan.DefaultMainDiskSizeBytes, cfg.MainDiskSizeBytes);
        Assert.Equal(VMNetworkMode.Nat, cfg.NetworkMode);
    }

    // ---- Serial console ⇔ debug grant (the hard guardrail) ----

    [Fact]
    public void ProductionRecipeGetsNoSerialConsole()
    {
        var cfg = VMConfig.Plan(MakeRecipe(), Json("{\"version\":2}"), Host16);
        Assert.False(cfg.SerialConsoleEnabled);
    }

    [Fact]
    public void DebugGrantSiblingEnablesTheSerialConsole()
    {
        var raw = Json("{\"version\":2,\"debugGrant\":\"{\\\"grant\\\":{},\\\"signatureHex\\\":\\\"ab\\\"}\"}");
        var cfg = VMConfig.Plan(MakeRecipe(), raw, Host16);
        Assert.True(cfg.SerialConsoleEnabled);
    }

    [Fact]
    public void DebugGrantInsideTheEnvelopeShapeEnablesTheConsole()
    {
        // The sibling rides at the TOP level of the issued envelope, beside
        // blob/blobSignature — exactly where the canonical engine reads it.
        var raw = Json("{\"blob\":{\"version\":2},\"blobSignature\":\"f1\",\"debugGrant\":{\"grant\":{\"issuedAt\":1},\"signatureHex\":\"ab\"}}");
        var cfg = VMConfig.Plan(MakeRecipe(), raw, Host16);
        Assert.True(cfg.SerialConsoleEnabled);
    }

    [Fact]
    public void EmptyDebugGrantStringDoesNotEnableTheConsole()
    {
        // Mirrors the engine's asStr: an empty string is "absent".
        var raw = Json("{\"version\":2,\"debugGrant\":\"\"}");
        var cfg = VMConfig.Plan(MakeRecipe(), raw, Host16);
        Assert.False(cfg.SerialConsoleEnabled);
    }

    // ---- Unlock policy from the signed blob ----

    [Fact]
    public void EncryptedGuestAwaitsPhoneUnlockAtBoot()
    {
        // Absent diskEncryption ⇒ LUKS ⇒ boots into the sealed state.
        var cfg = VMConfig.Plan(MakeRecipe(), Json("{}"), Host16);
        Assert.True(cfg.DiskEncrypted);
        Assert.True(cfg.AwaitsPhoneUnlockAtBoot);
        Assert.Equal("auto", cfg.BootUnlockMode);
    }

    [Fact]
    public void ApproveModeIsCarriedThrough()
    {
        var cfg = VMConfig.Plan(MakeRecipe(bootUnlockMode: "approve"), Json("{}"), Host16);
        Assert.Equal("approve", cfg.BootUnlockMode);
        Assert.True(cfg.AwaitsPhoneUnlockAtBoot);
    }

    [Fact]
    public void UnencryptedGuestBootsStraightThrough()
    {
        var cfg = VMConfig.Plan(MakeRecipe(diskEncryption: "none"), Json("{}"), Host16);
        Assert.False(cfg.DiskEncrypted);
        Assert.False(cfg.AwaitsPhoneUnlockAtBoot);
    }

    // ---- JSON round-trip (persisted in config.json) ----

    [Fact]
    public void JsonRoundTrip()
    {
        var raw = Json("{\"debugGrant\":\"x\"}");
        var cfg = VMConfig.Plan(MakeRecipe(bootUnlockMode: "approve"), raw, Host16);
        var json = JsonSerializer.Serialize(cfg, VMInventoryStore.SerializerOptions);
        var back = JsonSerializer.Deserialize<VMConfig>(json, VMInventoryStore.SerializerOptions);
        Assert.Equal(cfg, back);
        // The network mode serializes as the shared raw value.
        Assert.Contains("\"networkMode\": \"nat\"", json);
    }
}
